### PR TITLE
Client improvements and refactor

### PR DESCRIPTION
**Pull request by moowiz**
The client now keep track of policy names internally. Callers should not have to deal with any details of this.

This PR also refactors the client. It:

* Simplifies the client implementation
* Removes a few unnecessary methods
* Makes it have proper python types
* Adds docstrings to methods
* Adds basic tests for the client

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -10,6 +10,7 @@ show-source = True
 statistics = False
 doctests = True
 inline-quotes = double
+max-line-length = 88 # The black formatter expects a line length of 88 characters
```
Comment (moowiz) [inline setup.cfg]: I found this useful when developing locally. I can remove this if you want.

**Comment by moowiz** [comment]
This PR breaks the existing tests. I don't exactly understand how they work. They use environment variables. Should I try to fix them? Any thoughts?

File: setup.cfg (hunk 1 above)
Comment (bittermandel) [reply setup.cfg]: I think we should do either 79 (according to PEP8) or 119 (According to IntelliJ/other IDEs). I'm happy with either!

File: setup.cfg [hunk 2]
```diff
@@ -73,7 +73,6 @@ addopts =
 # mypy configurations: http://bit.ly/2zEl9WI
 allow_redefinition = False
 check_untyped_defs = True
-disallow_any_explicit = True
```
Comment (bittermandel) [inline setup.cfg]: How come we remove this?

**Comment by bittermandel** [comment]
> This PR breaks the existing tests. I don't exactly understand how they work. They use environment variables. Should I try to fix them? Any thoughts?

The existing tests are integration tests and I use them just to validate locally. We could mark them as such and only run on release/manually? What do you think?

File: setup.cfg (hunk 2 above)
Comment (moowiz) [reply setup.cfg]: I had thought I needed this for some typing, but it turns out I can just omit the type that's hard to represent. Added back.

**Comment by moowiz** [comment]
Ok, this passes all tests, lints, and type checks on python 3.8 and 3.9. There was some unittesting stuff added in 3.8 which I use here. I could probably figure out how to use pytest to do this if needed. Do we need to support python 3.6 and 3.7?

**Comment by bittermandel** [comment]
For this library it would be good if we support 3.6+, so it’s useful for more people :) I can check later today what we could do to make it work.

File: poe_client/rate_limiter.py [hunk 3]
```diff
@@ -106,27 +112,27 @@ async def parse_headers(self, headers) -> None:
...
+                        int(hits),
+                        int(restriction),
                     )
 
+        return policy_name
+
     async def get_semaphore(self, policy_name: str) -> bool:
         """Get a semaphore to make a request."""
-        if len(self.policies) == 0:
+        if not len(self.policies):
```
Comment (bittermandel) [inline poe_client/rate_limiter.py]: could just be `if not` right? An empty dict should be falsey.

File: pyproject.toml [hunk 4]
```diff
@@ -4,7 +4,10 @@ build-backend = "poetry.core.masonry.api"
 
 
 [tool.nitpick]
-style = "https://raw.githubusercontent.com/wemake-services/wemake-python-styleguide/master/styles/nitpick-style-wemake.toml"
+style = [
+  "https://raw.githubusercontent.com/wemake-services/wemake-python-styleguide/master/styles/nitpick-style-wemake.toml",
+  "https://raw.githubusercontent.com/moowiz/workaround-nitpick-issue-418/main/black-flake8-compat-style.toml",
```
Comment (bittermandel) [inline pyproject.toml:10]: This is nice, I like that. Could probably put it under this repo?

**Review by bittermandel — APPROVED** [review]

**Comment by bittermandel** [comment]
Good stuff! I like how you get the policy name from the first request and populate it that way, instead of assuming policies. Makes it more future proof and fault tolerant.

File: poe_client/rate_limiter.py (hunk 3 above)
Comment (moowiz) [reply poe_client/rate_limiter.py]: Yup, done.

File: pyproject.toml (hunk 4 above)
Comment (moowiz) [reply pyproject.toml:10]: I work on windows, and there's a bug in nitpick where it doesn't work correctly as a local file on windows. https://github.com/andreoliwa/nitpick/issues/418

Comment (moowiz) [reply pyproject.toml:10]: We could commit it and then point to the github main of this repo, but I don't know if that gives us much benefit.

File: setup.cfg (hunk 1 above)
Comment (moowiz) [reply setup.cfg]: If we want to not use 88 , then we need to change our black config somehow. I'm going to leave this as is (this change isn't in the repo in this state anymore), and we can change it in another PR if you want.

**Comment by moowiz** [comment]
I'm going to merge this PR. We'd need to rewrite the tests to work on 3.6 up. I couldn't figure out an easy way to do that. I might not be using pytest right, I'm using the builtin unittest stuff for python which is what I'm used to. Happy to look into this more if you think it's important to.

**Pull request merged**